origin-js/src/resources/marketplace.js

```diff
@@ -206,28 +206,51 @@ export default class Marketplace {
     const offerIds = await this.resolver.getOfferIds(listingId, opts)
     if (opts.idsOnly) {
       return offerIds
-    } else {
-      const allOffers = await Promise.all(
-        offerIds.map(async offerId => {
-          try {
-            return await this.getOffer(offerId)
-          } catch(e) {
-            // TODO(John) - handle this error better. It's tricky b/c it happens in a map
-            // and we want to throw the error, but we don't want the whole getOffers() call to fail.
-            // We want it to return the offers that it was able to get but still let us know something failed.
-            console.error(
-              `Error getting offer data for offer ${
-                offerId
-              }: ${e}`
-            )
-            return null
-          }
-        })
-      )
+    }
+
+    const allOffers = await Promise.all(
+      offerIds.map(async offerId => {
+        try {
+          return await this.getOffer(offerId)
+        } catch(e) {
+          // TODO(John) - handle this error better. It's tricky b/c it happens in a map
+          // and we want to throw the error, but we don't want the whole getOffers() call to fail.
+          // We want it to return the offers that it was able to get but still let us know something failed.
+          console.error(
+            `Error getting offer data for offer ${
+              offerId
+            }: ${e}`
+          )
+          return null
+        }
+      })
+    )
 
-      // filter out invalid offers
-      return allOffers.filter(offer => Boolean(offer))
+    // Filter out invalid offers
+    const filteredOffers = allOffers.filter(offer => Boolean(offer))
+    const listing = opts.listing || await this.getListing(listingId)
+    if (listing.type !== 'unit') {
+      return filteredOffers
     }
+
+    // Filter out offers for which the units purchased exceeds the units
+    // available at the time of the offer.
+    //
+    // TODO: Handle edits of unitsAvailable.
+    // TODO: Determine whether it always makes sense to filter out listings
+    // based on available quantity.
+    let unitsAvailable = listing.unitsTotal
+    return Object.keys(filteredOffers).reduce((offers, offerId) => {
+      const offer = filteredOffers[offerId]
+      if (offer.unitsPurchased > unitsAvailable) {
+        return offers
+      }
+      if (offer.status !== 'created' && offer.status !== 'withdrawn') {
```

Comment:
Maybe add a TODO about handling disputes/refunds case. If the buyer returns the item(s) maybe that should replenish the inventory ?